server/src/test/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequestTests.java

```diff
@@ -0,0 +1,132 @@
+        request.threadPool(randomBoolean());
+        request.transport(randomBoolean());
+        request.http(randomBoolean());
+        request.plugins(randomBoolean());
+        request.ingest(randomBoolean());
+        request.indices(randomBoolean());
+        NodesInfoRequest deserializedRequest = roundTripRequest(request);
+        assertThat(request.settings(), equalTo(deserializedRequest.settings()));
+    }
+
+    /**
+     * Test that a newly constructed NodesInfoRequestObject requests all of the
+     * possible metrics defined in {@link NodesInfoRequest.Metrics}.
+     */
+    public void testNodesInfoRequestDefaults() {
+        NodesInfoRequest defaultNodesInfoRequest = new NodesInfoRequest(randomAlphaOfLength(8));
+        NodesInfoRequest allMetricsNodesInfoRequest = new NodesInfoRequest(randomAlphaOfLength(8));
+        allMetricsNodesInfoRequest.all();
+
+        assertRequestsEqual(defaultNodesInfoRequest, allMetricsNodesInfoRequest);
+    }
+
+    /**
+     * Test that the {@link NodesInfoRequest#all()} method sets all of the
+     * metrics to {@code true}.
+     */
+    public void testNodesInfoRequestAll() throws Exception {
+        NodesInfoRequest request = new NodesInfoRequest("node");
+        request.all();
+
+        assertTrue(request.settings());
+        assertTrue(request.os());
+        assertTrue(request.process());
+        assertTrue(request.jvm());
+        assertTrue(request.threadPool());
+        assertTrue(request.transport());
+        assertTrue(request.http());
+        assertTrue(request.plugins());
+        assertTrue(request.ingest());
+        assertTrue(request.indices());
+    }
+
+    /**
+     * Test that the {@link NodesInfoRequest#clear()} method sets all of the
+     * metrics to {@code false}.
+     */
+    public void testNodesInfoRequestClear() throws Exception {
+        NodesInfoRequest request = new NodesInfoRequest("node");
+        request.clear();
+
+        assertFalse(request.settings());
+        assertFalse(request.os());
+        assertFalse(request.process());
+        assertFalse(request.jvm());
+        assertFalse(request.threadPool());
+        assertFalse(request.transport());
+        assertFalse(request.http());
+        assertFalse(request.plugins());
+        assertFalse(request.ingest());
+        assertFalse(request.indices());
+    }
+
+    /**
+     * Serialize and deserialize a request.
+     * @param request A request to serialize.
+     * @return The deserialized, "round-tripped" request.
+     */
+    private static NodesInfoRequest roundTripRequest(NodesInfoRequest request) throws Exception {
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            request.writeTo(out);
+            try (StreamInput in = out.bytes().streamInput()) {
+                return new NodesInfoRequest(in);
+            }
+        }
+    }
+
+    private static void assertRequestsEqual(NodesInfoRequest request1, NodesInfoRequest request2) {
```

Comment:
What do you think about exposing `requestedMetrics` but readonly? e.g.

```java
Set<String> getRequestedMetrics() {
   return Collections.unmodifiableSet(requestedMetrics);
}
```

Then the test can just do:

```java
assertThat(request1.getRequestedMetrics(), equalTo(request2.getRequestedMetrics()));
```

Reply:
I will do something like that on the next PR, when I change how this class interacts with other classes. In this refactoring PR, I wanted to focus on comparisons for the existing API. But you're right about where this should be heading.